src/core/lib/surface/channel_init.cc

```diff
@@ -20,34 +20,323 @@
 
 #include "src/core/lib/surface/channel_init.h"
 
+#include <string.h>
+
 #include <algorithm>
+#include <map>
+#include <set>
+#include <string>
+#include <type_traits>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/string_view.h"
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/channel/channel_stack_trace.h"
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/gprpp/crash.h"
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/surface/channel_stack_type.h"
 
 namespace grpc_core {
 
-void ChannelInit::Builder::RegisterStage(grpc_channel_stack_type type,
-                                         int priority, Stage stage) {
-  slots_[type].emplace_back(std::move(stage), priority);
+const char* (*NameFromChannelFilter)(const grpc_channel_filter*);
+
+namespace {
+struct CompareChannelFiltersByName {
+  bool operator()(const grpc_channel_filter* a,
+                  const grpc_channel_filter* b) const {
+    return strcmp(NameFromChannelFilter(a), NameFromChannelFilter(b)) < 0;
+  }
+};
+}  // namespace
+
+ChannelInit::FilterRegistration& ChannelInit::FilterRegistration::After(
+    std::initializer_list<const grpc_channel_filter*> filters) {
+  for (auto filter : filters) {
+    after_.push_back(filter);
+  }
+  return *this;
 }
 
+ChannelInit::FilterRegistration& ChannelInit::FilterRegistration::Before(
+    std::initializer_list<const grpc_channel_filter*> filters) {
+  for (auto filter : filters) {
+    before_.push_back(filter);
+  }
+  return *this;
+}
+
+ChannelInit::FilterRegistration& ChannelInit::FilterRegistration::If(
+    InclusionPredicate predicate) {
+  predicates_.emplace_back(std::move(predicate));
+  return *this;
+}
+
+ChannelInit::FilterRegistration& ChannelInit::Builder::RegisterFilter(
+    grpc_channel_stack_type type, const grpc_channel_filter* filter,
+    SourceLocation registration_source) {
+  filters_[type].emplace_back(
+      std::make_unique<FilterRegistration>(filter, registration_source));
+  return *filters_[type].back();
+}
+
+ChannelInit::StackConfig ChannelInit::BuildStackConfig(
+    const std::vector<std::unique_ptr<ChannelInit::FilterRegistration>>&
+        registrations,
+    PostProcessor* post_processors, grpc_channel_stack_type type) {
+  auto collapse_predicates =
+      [](std::vector<InclusionPredicate> predicates) -> InclusionPredicate {
+    switch (predicates.size()) {
```

Comment:
Do we need the optimization for the 0 and 1 cases here?  Seems like the general case would be fine even for those two values.

Reply:
worth doing the optimization if we're stuffing the result into a predicate itself: saves just a ton of indirect calls - but we don't need to, so lifted the loop out of the installed total predicate and into `struct Filter`.